File: src/main/java/de/pbauerochse/worklogviewer/youtrack/v20173/Grouping.java
package de.pbauerochse.worklogviewer.youtrack.v20173;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.pbauerochse.worklogviewer.youtrack.domain.GroupByCategory;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Grouping implements GroupByCategory {

    private final String id;
    private final String name;

    @JsonCreator
    public Grouping(@JsonProperty("id") String id, @JsonProperty("name") String name) {
        this.id = id;
        this.name = name;
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    @Override
    public boolean isValidYouTrackCategory() {
        return true;
    }
}

File: connectors/v20182/src/main/java/de/pbauerochse/worklogviewer/connector/v2018/domain/grouping/Grouping.kt
package de.pbauerochse.worklogviewer.connector.v2018.domain.grouping

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import de.pbauerochse.worklogviewer.connector.GroupByParameter
import de.pbauerochse.worklogviewer.connector.v2018.domain.issue.YouTrackIssue
import de.pbauerochse.worklogviewer.connector.v2018.domain.issue.YouTrackWorklogItem
import de.pbauerochse.worklogviewer.report.Issue

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "\$type")
@JsonSubTypes(
    JsonSubTypes.Type(value = FieldBasedGrouping::class, name = "jetbrains.youtrack.reports.impl.time.gap.FieldBasedGrouping")
)
interface Grouping : GroupByParameter {

    /**
     * Determines the grouping key for the given YouTrackWorklogItem
     * as of version 2018.2 YouTrack will not return the
     * grouping criteria. It has to be implemented in this application
     */
    fun getGroupingKey(
        worklogItem: YouTrackWorklogItem,
        issue: Issue,
        youtrackIssue: YouTrackIssue
    ): String?

}